client/rabbitmq.go

```diff
@@ -97,91 +133,90 @@ func (c *RabbitClient) Call(ctx context.Context, serviceName, endpoint string, r
 
 	// Don't even try to send if not connected
 	if !c.connection.IsConnected() {
-		return errors.Wrap(fmt.Errorf("Not connected to AMQP"))
+		return nil, errors.Wrap(fmt.Errorf("Not connected to AMQP"))
 	}
 
-	routingKey := c.buildRoutingKey(serviceName, endpoint)
+	replyChannel := c.inflight.push(req.Id())
 
-	correlation, err := uuid.NewV4()
-	if err != nil {
-		log.Errorf("[Client] Failed to create unique request id: %v", err)
-		return errors.Wrap(err) // @todo custom error code
-	}
-
-	log.Debugf("[Client] Dispatching request to %s with correlation ID %s", routingKey, correlation.String())
-
-	replyChannel := c.inflight.push(correlation.String())
-
-	requestBody, err := proto.Marshal(req)
-	if err != nil {
-		log.Errorf("[Client] Failed to marshal request: %v", err)
-		return errors.Wrap(err) // @todo custom error code
-	}
+	routingKey := buildRoutingKey(req.Service(), req.Endpoint())
+	log.Debugf("[Client] Dispatching request to %s with correlation ID %s", routingKey, req.Id())
 
+	// Build message from request
 	message := amqp.Publishing{
-		CorrelationId: correlation.String(),
+		CorrelationId: req.Id(),
 		Timestamp:     time.Now().UTC(),
-		Body:          requestBody,
+		Body:          req.Payload(),
 		ReplyTo:       c.replyTo,
+		Headers: amqp.Table{
+			"Content-Type":     req.ContentType(),
+			"Content-Encoding": "request",
+			"Service":          req.Service(),
+			"Endpoint":         req.Endpoint(),
+		},
 	}
 
-	err = c.connection.Publish(rabbit.Exchange, routingKey, message)
+	err := c.connection.Publish(rabbit.Exchange, routingKey, message)
 	if err != nil {
-		log.Errorf("[Client] Failed to publish %s to '%s': %v", correlation.String(), routingKey, err)
-		return errors.Wrap(err) // @todo custom error code
+		log.Errorf("[Client] Failed to publish %s to '%s': %v", req.Id(), routingKey, err)
+		return nil, errors.Wrap(err) // @todo custom error code
```

Comment:
Super minor issue, but I think this return may cause a tiny leak in the inflight registry